crates/valence_command/src/parsers/mod.rs

```diff
@@ -0,0 +1,278 @@
+        self.input.chars().nth(self.cursor)
+    }
+
+    pub fn peek_n(&self, n: usize) -> Option<char> {
+        self.input.chars().nth(self.cursor + n)
+    }
+
+    pub fn is_done(&self) -> bool {
+        self.cursor >= self.input.len()
+    }
+
+    pub fn advance(&mut self) {
+        self.cursor += 1;
+    }
+
+    pub fn pop(&mut self) -> Option<char> {
+        let c = self.peek();
+        if c.is_some() {
+            self.advance();
+        }
+        c
+    }
+
+    pub fn pop_n(&mut self, n: usize) -> Option<String> {
+        let s = self.input[self.cursor..self.cursor + n].to_string();
+        self.advance_n(n);
+        Some(s)
+    }
+
+    pub fn pop_to_next(&mut self, c: char) -> Option<String> {
+        if let Some(pos) = self.input[self.cursor..].find(c) {
+            let s = self.input[self.cursor..self.cursor + pos].to_string();
+            self.advance_n(pos);
+            Some(s)
+        } else {
+            None
+        }
+    }
+
+    pub fn pop_to_next_whitespace_or_end(&mut self) -> Option<String> {
+        match self.pop_to_next(' ') {
+            Some(s) => Some(s),
+            None => {
+                let s = self.input[self.cursor..].to_string();
+                self.advance_to(self.input.len());
+                Some(s)
+            }
+        }
+    }
+
+    pub fn match_next(&mut self, string: &str) -> bool {
+        if self.input[self.cursor..].to_lowercase().starts_with(string) {
+            self.advance_n(string.len());
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn skip_whitespace(&mut self) {
+        while let Some(c) = self.peek() {
+            if c.is_whitespace() {
+                self.advance();
+            } else {
+                break;
+            }
+        }
+    }
+
+    pub fn advance_n(&mut self, n: usize) {
+        self.cursor += n;
+    }
+
+    pub fn advance_to(&mut self, to: usize) {
+        self.cursor = to;
+    }
+
+    pub fn advance_to_next(&mut self, c: char) {
+        if let Some(pos) = self.input[self.cursor..].find(c) {
+            self.cursor += pos;
+        } else {
+            self.cursor = self.input.len();
+        }
+    }
+
+    pub fn advance_to_next_whitespace(&mut self) {
+        if let Some(pos) = self.input[self.cursor..].find(char::is_whitespace) {
+            self.cursor += pos;
+        } else {
+            self.cursor = self.input.len();
+        }
+    }
+
+    pub fn advance_to_next_non_whitespace(&mut self) {
+        if let Some(pos) = self.input[self.cursor..].find(|c: char| !c.is_whitespace()) {
+            self.cursor += pos;
+        } else {
+            self.cursor = self.input.len();
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum CommandArgParseError {
+    // these should be player facing and not disclose internal information
+    #[error("invalid argument, expected {0} got {1}")] // e.g. "integer" number
+    InvalidArgument(String, String),
```

Comment:
I would prefer named fields to make the distinction between the strings clearer. Something like
```suggestion
    InvalidArgument {
        expected: String,
        got: String,
    },
```